kore/src/Kore/Attribute/Pattern/ConstructorLike.hs

```diff
@@ -193,15 +195,42 @@ A builtin value is not technically a constructor, but it is constructor-like for
 builtin domains, at least from the perspective of normalization (normalized
 means constructor-like here).
  -}
-instance Synthetic ConstructorLike (Builtin key) where
+instance HasConstructorLike key => Synthetic ConstructorLike (Builtin key)
+  where
     synthetic =
         \case
             BuiltinInt _    -> ConstructorLike . Just $ ConstructorLikeHead
             BuiltinBool _   -> ConstructorLike . Just $ ConstructorLikeHead
             BuiltinString _ -> ConstructorLike . Just $ ConstructorLikeHead
+            (BuiltinMap InternalAc
+                    {builtinAcChild = NormalizedMap builtinMapChild}
+                ) -> normalizedAcConstructorLike builtinMapChild
+            (BuiltinSet InternalAc
+                    {builtinAcChild = NormalizedSet builtinSetChild}
+                ) -> normalizedAcConstructorLike builtinSetChild
             _               -> ConstructorLike Nothing
     {-# INLINE synthetic #-}
 
+normalizedAcConstructorLike
+    ::  ( HasConstructorLike key
+        , HasConstructorLike (Value collection ConstructorLike)
+        )
+    => NormalizedAc collection key ConstructorLike -> ConstructorLike
+normalizedAcConstructorLike ac@(NormalizedAc _ _ _) =
+    case ac of
+        NormalizedAc
+            { elementsWithVariables = []
+            , concreteElements
+            , opaque = []
+            }
+              | all pairIsConstructorLike concreteElementsList
+                -> ConstructorLike . Just $ ConstructorLikeHead
+              where
+                concreteElementsList = Map.toList concreteElements
+                pairIsConstructorLike (key, value) =
+                    isConstructorLike key && isConstructorLike value
```

Comment:
Why do we need to check the `key`? It should always be constructor-like. If we are going to check, I think it should at least be an assertion, but if we don't check, then we also don't need the `HasConstructorLike` abstraction.

Reply:
We don't, but I thought it's safer to check it, especially since the requirements for the key changed in the past. You're right, an assertion makes more sense, so I changed it to that.